### PR TITLE
Final Implementations (Pair Programming)

### PR DESCRIPTION
**Pull request by pokecedgo**
Decided on doing the following:
DigitalRootBatchJobManager 
 DistributedDigitalRootCluster Implementations

**DigitalRootBatchJobManager:** Handles batch computation of digital roots for a list of integers.
**DistributedDigitalRootCluster:** Simulates distributed computation of digital roots using a thread pool.
These implementations provide both batch and distributed approaches for digital root calculations, improving flexibility and scalability for processing large dataset

Small features: Terminal visual display of the RootChain application for interactivity.

### CODE REVIEW

File: src/project/Main.java [hunk 1]
```diff
@@ -51,82 +73,100 @@ public static void main(String[] args) throws Exception {
                 f.readFile(fileName);
                 System.out.println("File read and computation complete.");
             } else {
-                System.out.println("Enter input separated by the spaces:");
+                System.out.println("Enter input separated by commas:");
```
Comment (Antonio-Vav) [inline src/project/Main.java:76]: catching this probably prevented a future bug

File: src/project/Main.java [hunk 2]
```diff
@@ -51,82 +73,100 @@ public static void main(String[] args) throws Exception {
...
-     * Tests the DigitalRootPersistenceAPI implementation.
+    /**
+     * Test DigitalRootPersistenceAPI
+     * Tests and benchmarks both implementations.
      */
     private static void testDigitalRootPersistenceAPI() {
         // Test DigitalRootPersistenceAPI
         ImplementDigitalRootPersistenceAPI api = new ImplementDigitalRootPersistenceAPI();
-        project.apis.computeapi.FastDigitalRootPersistenceAPI fastApi = new project.apis.computeapi.FastDigitalRootPersistenceAPI();
+        FastDigitalRootPersistenceAPI fastApi = new FastDigitalRootPersistenceAPI();
```
Comment (Antonio-Vav) [inline src/project/Main.java:112]: More efficient, I would ensure the code still functions as expected if not done already.

File: src/project/Main.java [hunk 3]
```diff
@@ -51,82 +73,100 @@ public static void main(String[] args) throws Exception {
...
-        // Benchmark original vs optimized
-        int largeNumber = 987654321;
-        int iterations = 1_000_000;
+        System.out.println("\nOptimized Implementation:");
+        for (int i = 0; i < testNumbers.length; i++) {
+            String result = fastApi.processDigitalRootPersistence(testNumbers[i]);
+            boolean correct = result.equals(expectedResults[i]);
+            System.out.println("Input: " + testNumbers[i] + ", Expected: " + expectedResults[i] + ", Actual: " + result + (correct ? " ✓" : " ✗"));
+        }
+        System.out.println("\nBenchmark Results:");
```
Comment (Antonio-Vav) [inline src/project/Main.java:132]: Good to test at meaningful capacities like this

File: src/main/proto/networkapi.proto [hunk 4]
```diff
@@ -8,19 +8,19 @@ service NetworkAPI {
 }
 
 message AskUserRequest {
-    required string inputPath = 1;
-    required string outputPath = 2;
+    optional string inputPath = 1;
+    optional string outputPath = 2;
```
Comment (WestwardSky8392) [inline src/main/proto/networkapi.proto:12]: Way to switch to optional!

File: ina/DigitalRootBatchJobManager.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package ina;
+
+import java.util.List;
+import java.util.ArrayList;
+
```
Comment (WestwardSky8392) [inline ina/DigitalRootBatchJobManager.java:5]: The DigitalRootBatchJobManager is an implementation for computing the digital roots of the list of integers. The code follows good organization, encapsulating the digital root logic within a private method for reuse.

File: ina/DigitalRootBatchJobManager.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+public class DigitalRootBatchJobManager {
+    public List<Integer> computeBatchDigitalRoots(List<Integer> numbers) {
+        List<Integer> results = new ArrayList<>();
+        for (Integer n : numbers) {
+            results.add(digitalRoot(n));
+        }
+        return results;
+    }
+
+    private int digitalRoot(int n) {
```
Comment (WestwardSky8392) [inline ina/DigitalRootBatchJobManager.java:15]: The digital root computation is implemented using a loop-based approach, which is easy to understand and avoids recursion.

File: ina/DistributedDigitalRootCluster.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.List;
+import java.util.concurrent.*;
+
+public class DistributedDigitalRootCluster {
+    private final ExecutorService executor;
+
+    public DistributedDigitalRootCluster(int nodeCount) {
+        this.executor = Executors.newFixedThreadPool(nodeCount);
+    }
+
```
Comment (WestwardSky8392) [inline ina/DistributedDigitalRootCluster.java:12]: This code on the other hand from the pervious file, submits each new task into the thread pool. Allowing for better runtime and more efficient work.

File: src/project/Main.java [hunk 8]
```diff
@@ -51,82 +73,100 @@ public static void main(String[] args) throws Exception {
...
         int[] testNumbers = {9875, 12345, 999, 0, 1, 77};
         String[] expectedResults = {"2", "6", "9", "0", "1", "5"};
 
         System.out.println("Testing DigitalRootPersistenceAPI:");
+        System.out.println("================================");
+        System.out.println("Original Implementation:");
         for (int i = 0; i < testNumbers.length; i++) {
             String result = api.processDigitalRootPersistence(testNumbers[i]);
-            System.out.println("Input: " + testNumbers[i] + ", Expected: " + expectedResults[i] + ", Actual: " + result);
+            boolean correct = result.equals(expectedResults[i]);
```
Comment (Antonio-Vav) [inline src/project/Main.java:123]: This is definitely more effective of a test since it takes in many different values

File: src/project/apis/computeapi/FastDigitalRootPersistenceAPI.java [hunk 9]
```diff
@@ -1,30 +1,47 @@
...
-            steps++;
         }
-        // Return the digital root as a string (to match interface)
         return String.valueOf(n);
     }
+
+    /**
+     * Optimized digital root calculation for arbitrarily large numbers using BigInteger.
+     */
+    public String processDigitalRootPersistence(BigInteger number) {
```
Comment (Antonio-Vav) [inline src/project/apis/computeapi/FastDigitalRootPersistenceAPI.java:33]: Good to test edge cases like the value being null
